### ISSUE TITLE
Can't backup to the S3 - fails after starting

### ISSUE THREAD

**Issue by adkinsn** [pgbackrest/pgbackrest#2200]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version:
      - v2.48
2. PostgreSQL version:
     - v12.8
3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
Each system has Red Hat Enterprise Linux Server release 7.9 (Maipo).  These are dedicated database servers.  Two are virtual machines  and the other one is a physical server serving as the primary node out of a three-node cluster.

5. Did you install pgBackRest from source or from a package?
`yum install pgbackrest`
`dzdo su postgres`

6. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
              - attached
    - `postgresql.conf` settings applicable to pgBackRest:
              - max_wal_senders = 2
              - wal_level = replica
              - archive_mode = on
              - archive_command = 'pgbackrest --stanza=main archive-push %p'
              - listen_addresses = 'localhost,127.0.0.1,<redacted system address>,<redacted db alias>'
              - port = 5432
    - errors in the postgresql log file before or during the time you experienced the issue
          - attached
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

8. Describe the issue:
We are having a number of issues with backups failing on multiple systems.  Over the past few weeks, there were recovery efforts and updates to the S3 lifecycle policies.  I am not sure if they are related or not as the dates of the last backups vary across all three of our Postgres environments.  It looks like the files will start writing to the S3 then it fails.  We are not finding anything in the logs.  We have involved the AWS Engineers and SysAdmins for the local systems along with our DBA team and we are unable to figure out what is causing this.  Could it be conflicts in the S3 policy (which are now disabled)?  Could it be the lack of prior backup it can reference to drop?  Our attempts to rerun it are about 6 minutes before it fails.  For one of the systems, we Shut down and restarted our production environment and changed the archive mode.  Updated the mode and restarted the service.  We reran the PGBACKREST full backup command.  It failed with the same timeout error.  We could really use some assistance on this one as we are at a loss on what could be causing this.  It looks like pgbackrest has been updated in September to the latest version, so I am not sure how or if at all plays into this. 

These are critical databases to our organization, so any assistance or guidance you can provide would be helpful.  I am happy to schedule a support call if observation is required.
[postgresql_beforeafter_pgbackrest.log](https://github.com/pgbackrest/pgbackrest/files/12834892/postgresql_beforeafter_pgbackrest.log)
[pgbackrest_Stage_Backup_ErrorLog.txt](https://github.com/pgbackrest/pgbackrest/files/12834894/pgbackrest_Stage_Backup_ErrorLog.txt)

**Comment by dwsteele** [comment]
The log fragment you provided is too short to be sure, but it looks like pgBackRest is having an issue overwriting `backup.manifest.copy`. This may be due to policy changes in S3? You can disable this write by setting `resume=n` but other files get overwritten later that are not optional, e.g. `backup.info`.

>  It looks like pgbackrest has been updated in September to the latest version, so I am not sure how or if at all plays into this

This doesn't seem too likely but it is possible. You could try reverting if nothing else works.

**Comment by adkinsn** [comment]
@dwsteele - We tested a theory.  We downgraded to 2.47 and it worked. with the exception of the issue below.  So it seems the upgrade may have something to fail.  I guess my remaining question is the errors below.  How can I properly address them. Is there a way to “reset” the series?  

Our backup schedule is one weekly full backup with differential backups for the remaining days.  There are S3 lifecycle policies that kick in where backup files move into deep archive after 21 days.  Files get dropped after 90 days with the exception of the weekly full backups which drop after 3 years.  

The challenge is that we have with the “indefinite” option below is that it may result in additional issues.

What are your thoughts?  


HINT: to retain full backups indefinitely (without warning), set option 'repo1-retention-full' to the maximum.
2023-10-10 14:31:25.255 P00   INFO: execute non-exclusive backup start: backup begins after the next regular checkpoint completes
2023-10-10 14:31:25.756 P00   INFO: backup start archive = 0000000500007B940000001B, lsn = 7B94/1B000028
2023-10-10 14:31:25.756 P00   INFO: check archive for prior segment 0000000500007B940000001A
WARN: backup '20231005-162932F' cannot be resumed: new pgBackRest version '2.47' does not match resumable pgBackRest version '2.48'
2023-10-10 18:31:49.833 P00   INFO: execute non-exclusive backup stop and wait for all WAL segments to archive
2023-10-10 18:31:50.033 P00   INFO: backup stop archive = 0000000500007B940000005A, lsn = 7B94/5AB39578
2023-10-10 18:31:50.274 P00   INFO: check archive for segment(s) 0000000500007B940000001B:0000000500007B940000005A
2023-10-10 18:31:53.176 P00   INFO: new backup label = 20231010-143125F
2023-10-10 18:31:55.117 P00   INFO: full backup size = 4585.8GB, file total = 19411
2023-10-10 18:31:55.119 P00   INFO: backup command end: completed successfully (14431503ms)
2023-10-10 18:31:55.119 P00   INFO: expire command begin 2.47: --exec-id=11883-146ef522 --log-level-console=info --repo1-path=/db-bkups/pgbackrest/<redacted>/ --repo1-s3-bucket=<redacted>]--repo1-s3-endpoint=s3.us-east-1.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-east-1 --repo1-type=s3 --stanza=main
2023-10-10 18:31:55.924 P00   INFO: option 'repo1-retention-archive' is not set - archive logs will not be expired
2023-10-10 18:31:55.924 P00   INFO: expire command end: completed successfully (805ms)

**Comment by dwsteele** [comment]
Did you follow the hint?
```
HINT: to retain full backups indefinitely (without warning), set option 'repo1-retention-full' to the maximum.
```
https://pgbackrest.org/configuration.html#section-repository/option-repo-retention-full

**Comment by dwsteele** [comment]
There's not enough info in this log to do any debugging. Please provide the entire backup log with `--log-level-console=trace`.

**Comment by dwsteele** [comment]
You provided a log fragment but we need the entire log.

You could also try increasing `io-timeout`. The only change I can see in 2.48 that is even vaguely related here is we made connection timeouts stricter, i.e. more closely to what `io-timeout` says. I can't see how it relates here, but it is possible.

**Comment by tanelsuurhans** [comment]
Don't have logs to provide just yet, but we are seeing the same exact issue with backups to S3. We tried increasing `io-timeout` to no avail. Backups consistently fail on 2.48 at the same spot across 5 different databases and both `full` and `diff` backups. The only thing that fixed this was downgrading to 2.47. Hopefully we'll have time today to file a ticket with more information.

**Comment by dwsteele** [comment]
We have not been able to reproduce this issue, but we are pretty sure it is due to a regression in how retries are handled. See #2221. Closing since there is now a PR for this issue.

@adkinsn How would you like your name to appear in the release notes?

**Comment by adkinsn** [comment]
Please use “Norman Adkins” in the release notes.

**Comment by adkinsn** [comment]
> We have not been able to reproduce this issue, but we are pretty sure it is due to a regression in how retries are handled. See #2221. Closing since there is now a PR for this issue.
> 
> @adkinsn How would you like your name to appear in the release notes?

@dwsteele - you can use Norman Adkins.  Thanks.

**Comment by Deltahedron** [comment]
I see that the fix was already merged and closed but I'll add my findings.

AWS Ubuntu 22.04.3:
- the old version 2.47-1.pgdg22.04+1 works on machines that haven't been upgraded and reverting to that version on other machines that have been upgraded fixed the issue
- the new version2.48-1.pgdg22.04+1 failed with the S3 timeout

Azure RHEL 8.8:
- the new version 2.48-1PGDG.rhel8 works normally

**Comment by dwsteele** [comment]
@Deltahedron We have also found that the issue is kernel specific. It started happening a few years ago after a round of kernel updates -- reverting the kernel fixed the problem then. But of course that was not a long term solution for us so we implemented non-blocking TLS so we could at least timeout and try again when the connection went bad. The blocking TLS write always did eventually return, but it could be 15-20 minutes.

So really, we have the same underlying problem -- TLS writes sometimes stall. We have retries to deal with this but still have no idea why the stalls are happening.

We also use TLS to communication between pgBackRest servers and we have never seen any stalls there. Client code is the same with no retries. So maybe it is something weird S3 is doing but I'd sure like to know what.
